core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java

```diff
@@ -0,0 +1,287 @@
+ *
+ * <p><b>Local datacenter</b>: This implementation will only define a local datacenter if it is
+ * explicitly set either through configuration or programmatically; if the local datacenter is
+ * unspecified, this implementation will effectively act as a datacenter-agnostic load balancing
+ * policy and will consider all nodes in the cluster when creating query plans, regardless of their
+ * datacenter.
+ *
+ * <p><b>Query plan</b>: This implementation prioritizes replica nodes over non-replica ones; if
+ * more than one replica is available, the replicas will be shuffled. Non-replica nodes will be
+ * included in a round-robin fashion. If the local datacenter is defined (see above), query plans
+ * will only include local nodes, never remote ones; if it is unspecified however, query plans may
+ * contain nodes from different datacenters.
+ *
+ * <p><b>This class is not recommended for normal users who should always prefer {@link
+ * DefaultLoadBalancingPolicy}</b>.
+ */
+@ThreadSafe
+public class BasicLoadBalancingPolicy implements LoadBalancingPolicy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BasicLoadBalancingPolicy.class);
+
+  protected static final IntUnaryOperator INCREMENT = i -> (i == Integer.MAX_VALUE) ? 0 : i + 1;
+
+  @NonNull protected final InternalDriverContext context;
+  @NonNull protected final DriverExecutionProfile profile;
+  @NonNull protected final String logPrefix;
+
+  protected final AtomicInteger roundRobinAmount = new AtomicInteger();
+  protected final CopyOnWriteArraySet<Node> liveNodes = new CopyOnWriteArraySet<>();
+
+  // private because they should be set in init() and never be modified after
+  private volatile DistanceReporter distanceReporter;
+  private volatile Predicate<Node> filter;
+  private volatile String localDc;
+
+  public BasicLoadBalancingPolicy(@NonNull DriverContext context, @NonNull String profileName) {
+    this.context = (InternalDriverContext) context;
+    profile = context.getConfig().getProfile(profileName);
+    logPrefix = context.getSessionName() + "|" + profileName;
+  }
+
+  /** @return The local datacenter, if known; empty otherwise. */
+  public Optional<String> getLocalDatacenter() {
+    return Optional.ofNullable(localDc);
+  }
+
+  /**
+   * @return An immutable copy of the nodes currently considered as live; if the local datacenter is
+   *     known, this set will contain only nodes belonging to that datacenter.
+   */
+  public Set<Node> getLiveNodes() {
+    return ImmutableSet.copyOf(liveNodes);
+  }
+
+  @Override
+  public void init(@NonNull Map<UUID, Node> nodes, @NonNull DistanceReporter distanceReporter) {
+    this.distanceReporter = distanceReporter;
+    localDc = newLocalDcHelper().discoverLocalDc().orElse(null);
+    filter = newNodeFilterHelper().createNodeFilter(localDc);
+    for (Node node : nodes.values()) {
+      if (filter.test(node)) {
+        distanceReporter.setDistance(node, NodeDistance.LOCAL);
```

Comment:
Is it safe to set distance to `LOCAL`even if the node filter works for BasicStrategy that doesn't require dc?

Reply:
I think so. At least this mimics the behavior of the old driver `com.datastax.driver.core.policies.RoundRobinPolicy`, which always returns `LOCAL` for all nodes in the cluster.